olp-cpp-sdk-core/src/network2/ios/OLPNetworkIOS.mm

```diff
@@ -0,0 +1,366 @@
+    headers[key] = value;
+  }
+  return headers;
+}
+
+NSData* ParseBodyDataFromRequest(const olp::network2::NetworkRequest& request) {
+  const auto verb = request.GetVerb();
+  if (verb == NetworkRequest::HttpVerb::GET ||
+      verb == NetworkRequest::HttpVerb::HEAD) {
+    return nil;
+  }
+
+  const auto body = request.GetBody();
+  if (!body || body->empty()) {
+    return nil;
+  }
+  return [NSData dataWithBytes:static_cast<const void*>(body->data())
+                        length:body->size()];
+}
+
+NSString* ParseHttpMethodFromRequest(
+    const olp::network2::NetworkRequest& request) {
+  switch (request.GetVerb()) {
+    case NetworkRequest::HttpVerb::GET:
+      return OLPHttpMethodGet;
+    case NetworkRequest::HttpVerb::POST:
+      return OLPHttpMethodPost;
+    case NetworkRequest::HttpVerb::PUT:
+      return OLPHttpMethodPut;
+    case NetworkRequest::HttpVerb::DEL:
+      return OLPHttpMethodDelete;
+    case NetworkRequest::HttpVerb::PATCH:
+      return OLPHttpMethodPatch;
+    case NetworkRequest::HttpVerb::HEAD:
+      return OLPHttpMethodHead;
+    default:
+      return nil;
+  }
+}
+
+olp::network2::ErrorCode ConvertNSURLErrorToNetworkErrorCode(
+    NSInteger error_code) {
+  switch (error_code) {
+    case NSURLErrorUnsupportedURL:
+    case NSURLErrorCannotFindHost:
+      return ErrorCode::INVALID_URL_ERROR;
+    case NSURLErrorNotConnectedToInternet:
+      return ErrorCode::OFFLINE_ERROR;
+    case NSURLErrorTimedOut:
+      return ErrorCode::TIMEOUT_ERROR;
+    default:
+      if (error_code >= NSURLErrorClientCertificateRequired &&
+          error_code <= NSURLErrorSecureConnectionFailed) {
+        return ErrorCode::AUTHORIZATION_ERROR;
+      } else {
+        return ErrorCode::UNKNOWN_ERROR;
+      }
+      break;
+  }
+}
+}  // namespace
+
+#pragma mark - OLPNetworkIOS implementation
+
+OLPNetworkIOS::OLPNetworkIOS(size_t max_requests_count)
+    : max_requests_count_(max_requests_count) {
+  @autoreleasepool {
+    http_client_ = [[OLPHttpClient alloc] init];
+  }
+}
+
+OLPNetworkIOS::~OLPNetworkIOS() { Cleanup(); }
+
+olp::network2::NetworkStatus OLPNetworkIOS::Send(
+    olp::network2::NetworkRequest request,
+    std::shared_ptr<std::ostream> payload,
+    olp::network2::Network::Callback callback,
+    olp::network2::Network::HeaderCallback header_callback,
+    olp::network2::Network::DataCallback data_callback) {
+  @autoreleasepool {
+    OLPHttpTask* task = nil;
+    NSString* url = [NSString stringWithUTF8String:request.GetUrl().c_str()];
+    if (url.length == 0) {
+      EDGE_SDK_LOG_WARNING(kLogTag, "Invalid request URL!");
+      return NetworkStatus(ErrorCode::INVALID_URL_ERROR);
+    }
+
+    // create new OLPHttpTask, if possible
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (http_client_.activeTasks.count >= max_requests_count_) {
+        EDGE_SDK_LOG_WARNING_F(
+            kLogTag,
+            "Can't send request:[%s] - reached max requests count:[%lu / %lu]!",
```

Comment:
Do we need to print URL. Might get pretty big, maybe move at the end:

"Max request count exceeded, URL=[%s]"
I also dont think we need punctuation marks at the end. Warnings should be pretty clear, no need to add additional '!' at the end